README.md
# README #

## DehashedDomains ##

**Version 1.0.0**

### Purpose ###

This is a single Python script that interfaces with Dehashed's (https://dehashed.com/) API to quickly scour the dark web for breached company data given a user-supplied domain. 

### Setup ###

Prior to running the application, ensure that you have set your Dehashed login email and API key in your environment variables.

They need to be named "DEHASH_EMAIL" and "DEHASH_API" respectfully.

In Linux, add these environment variables by editing the following file as administrator:

```bash
If using 'vim' as an editor:

sudo vim /etc/environment

```

NOTE: After doing so, you will either need to logout/log back in or reboot your machine for the effects to take place. Simply running 'source /etc/environment' will not suffice as it is not a script.

ALSO: Please install 'name-that-hash' (more info found here: https://www.kali.org/tools/name-that-hash/) with apt to utilize the hash-identifier/organizer features like so : 

```bash
sudo apt install name-that-hash

```

After you do this, your script will be ready to run!

As stated before, you will need to provide a login email and API key both relevant to your *paid* Dehashed.com account as environment variables.

NOTE: If you or anyone else generates a new API key for any reason (which can easily be done with the push of a button in your Dehashed.com account settings), it will render all old API keys invalid.

### Usage ###

The script doesn't require any outside dependencies to run, just the newest version of Python. However, to have it automatically try to identify any returned hash types, install 'name-that-hash' as stated above. 

Upon running dehash.py, you must specify a single domain that you wish to gather information on with '--domain'

It will then attempt to:

Authenticate to your Dehashed.com account

Make a single API call to Dehashed that returns the first 10,000 results from Dehashed.com results on given domain

Create a new dynamically named folder in the current directory based on the provided domain that contains a:

* CSV file with ALL results

* Text file of usernames, one per line

* Text file of passwords, one per line

* Text file of credentials, (username:password) one per line (if this file is empty, there were no results that contained BOTH a username and a password)

* Text file of hashed passwords, one per line (alphabetically sorted)

* Text file of user emails, one per line (removes duplicates)

* Folder with files that are sorted by hash type (as accurate as the name-that-hash's Python module can be) and named according to Hashcat's mode number (specific to hash type).

NOTE: Each time the script is run, it *should* only consume 1 API token from Dehashed.com

### Cracking Results with Hashcat ###  

This script generates a folder named "hashcat-modes" that sorts all identified hashes by hash type. Once sorted by hash type, they are added to a text file named after the Hashcat mode associated with that hash type. 

This allows for easy cracking. For example, specify a text file named hashcat_mode_100 in the command like so (which correlates to the hash type SHA1 in hashcat): 

```bash
hashcat -a 0 -m 100 ./hashcat_mode_100 ~/Tools/rockyou.txt
```

### Example ###

```bash
python ./Documents/DehashedDomains/dehash.py --domain example.com

```

### Potential features to be added ###

Add redundant code to increase the 10,000 result limit to the API’s max of 30,000 (would increase the API token usage from 1 -> 3 per script run)

### Who do I talk to? ###

For any questions, comments, or other feedback please feel free to submit a pull request or ping me here on GitHub: https://github.com/jlanguell


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2023-09-10
### Added
- Initial repo creation
- Added LICENSE
- Added version.txt
- Added CHANGELOG.md
- Added README.md
- Added dehash.py
- Added .gitignore


dehash.py
# 09-10-23
# https://github.com/jlanguell

# This Python script accepts a single domain in CLI args and environment variables to authenticate with Dehashed API, gather breached data on client, and write it all to the user's home directory.
# Additionally, any found hashes are checked with name-that-hash application (must be installed via Linux) for their hashcat *mode*, and saved accordingly to help speed up hash-cracking methods.

# NOTE: Script requires a domain as input with --domain or -d
# NOTE: Need to 'sudo apt install name-that-hash' to utilize automatic hash identifier function


# Imports

import sys
import os
import json
import csv
import argparse
import requests
import subprocess
from requests.auth import HTTPBasicAuth


def parse_args(): # Parse CLI args
	# CLI args
	parser = argparse.ArgumentParser(description = 'Dehashed Domain Scraper\nScript gathers the first 10,000 data entries, which is the max amount for Dehashed API queries.')
	parser.add_argument('-d', '--domain', help = 'Please specify a domain to scan such as "google.com"')
	args = parser.parse_args(sys.argv[1:])
	domain = args.domain
	if not domain:
		print("Please specify a domain with -d")
		sys.exit(0)
	return domain
	
	
def get_env(): # Authorize API connection via local environment variables (these need to be defined & stored in /etc/environment for linux)
	try:
		username = os.environ.get("DEHASH_EMAIL")
		key = os.environ.get("DEHASH_API")
		return username, key;
	except:
		print('Please ensure that you have login email and API key stored as environment variables named "DEHASH_EMAIL" & "DEHASH_API". For linux, you can create these by editing /etc/environment with sudo privileges.')
		sys.exit(0)


def http_request(url,username,key,headers_dict):
	request = requests.get(url, auth = HTTPBasicAuth(username,key),headers=headers_dict)
	json_data = json.loads(request.text)
	try:
		entries = json_data['entries']
	except:
		print('There was an error connecting to API, please ensure API email and key are set in your environment variables as "DEHASH_EMAIL" and "DEHASH_API" respectfully.')
		sys.exit(0)
	if not entries:
		print('No entries found for that domain.')
		sys.exit(0)
	return entries


def create_dir(domain):
	# Create Domain Folder in User Home Directory
	home_directory = os.path.expanduser('~')
	path = os.path.join(home_directory, "%s-dehashed" %domain)
	path_exist = os.path.exists(path)
	if not path_exist:
		os.makedirs(path)
		return path
	else:
		print('Scan results for this domain already exist at %s.\n\nPlease "rm -rf" existing Dehashed results folder for this domain or run script in a different directory.' %path)
		sys.exit(0)


def create_csv(path,entries):
	csv_file = open('%s/all-data.csv' %path, 'w')
	csv_writer = csv.writer(csv_file)
	count=0
	for entry in entries:
		if count == 0:
			header = entry.keys()
			csv_writer.writerow(header)
			count+=1	
		csv_writer.writerow(entry.values())   
	csv_file.close()


def write_file(path,entries,name):
	list=[]
	with open('%s/%s.txt' %(path,name), 'w') as write_file:
		
		if name=="creds": # Write username:password credentials to file whenever both are present in a single entry
			for entry in entries: 	 
					if entry['username'] != "" and entry['password'] != "":
						write_file.write(entry['username'] + ":" + entry['password'] + "\n")
			return
		
		if name=="hashes": # Write hashes to file (sorted alphabetically)
			for entry in entries:
				if entry['hashed_password'] != "":
					list.append(entry['hashed_password'] + "\n")
			list.sort()
			for item in list:
				write_file.write(item)
			return
		
		if name=="emails": # Write emails to file (sorted alphabetically)
			for entry in entries:
				if entry['email'] != "":
					list.append(entry['email'].lower() + "\n")  
			list.sort()
			list = set(list)
			for item in list:
				write_file.write(item)
			return

		for entry in entries: # Write username or password to respective file
			if entry['%s' %name] != "": 
				write_file.write(entry['%s' %name] + "\n")


def identify_hash(path,domain): # Must 'sudo apt install name-that-hash' prior to running this program
	path_exist = os.path.exists("%s/hashcat-modes/" %path)
	if not path_exist:
		os.makedirs("%s/hashcat-modes/" %path)
	else:
		print("An error occurred. Please make sure you delete or move any old file folders for %s in your home directory." %domain)
		sys.exit(0)
	result = subprocess.run('nth -f %s/hashes.txt -g' %(path), universal_newlines=True, capture_output=True, shell=True)
	with open('%s/hashcat-modes/name_that_hash.json' %path, "w") as f: # Creates a file containing all JSON from name-that-hash command
		f.write(result.stdout)
	with open('%s/hashcat-modes/name_that_hash.json' %path, "r") as f: # Tries to load JSON from nth, or catches "nth cmd not found" json null-loading error
		try:
			returned = json.load(f, strict=False)
		except json.decoder.JSONDecodeError as e:
			print("ERROR: HASH-IDENTIFIER DID NOT RUN \n\nPlease make sure you have already ran 'sudo apt install name-that-hash' prior!")
			sys.exit(0)
	    
	for key, value in returned.items(): # writes each hash & hashcat mode to respective file
		try:
			if value[0]["hashcat"] != "" or None:        
				mode = value[0]["hashcat"]	    
			elif value[1]["hashcat"] != "" or None:
				mode = value[1]["hashcat"]
			elif value[2]["hashcat"] != "" or None:
				mode = value[2]["hashcat"]
			else:
				print(key + "Not Identified")
			with open("%s/hashcat-modes/hashcat_mode_%s" %(path,mode), "a") as f:
				f.write(key + "\n")
	
		except: # writes hash to this file if not identified by nth
			with open("%s/hashcat-modes/hashcat_mode_NOTFOUND" %path, "a") as f: 
				f.write(key + "\n")
		

def main(): # Main function calls all other functions
	headers_dict = {'Accept':'application/json', # Prebuilt header variable
		'User-Agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.3538.77 Safari/537.36'} # Can't use python default User-Agent
	domain = parse_args() # Parse command line arguments to get domain
	url = 'https://api.dehashed.com/search?query=domain:"%s"&size=10000' % (domain) # API query address; can alter size within limits (see Dehashed documentation on their website)	
	username, key = get_env() # Get Dehashed username and key from ENV variables
	entries = http_request(url,username,key,headers_dict) # Authenticate and request data from Dehashed
	path = create_dir(domain) # Creates a folder in user's home directory to store all created data
	create_csv(path,entries) # Writes all Dehashed data to all-data.csv
	write_file(path,entries,name="username") # Writes data to username.txt
	write_file(path,entries,name="password") # Writes data to password.txt
	write_file(path,entries,name="creds") # Writes data to creds.txt
	write_file(path,entries,name="hashes") # Writes data to hashes.txt
	write_file(path,entries,name="emails") # Writes data to emails.txt
	identify_hash(path,domain) # Identifies hashes via name-that-hash subprocess and categorizes them all by file in /path/hashcat-modes/
	print("If the operation was successful, the data was saved to: %s" %path)	
	
	
if __name__ == "__main__": # Run application
	main()
	
	
